RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc

```diff
@@ -0,0 +1,632 @@
+#include "RecoLocalCalo/HcalRecAlgos/interface/MahiFit.h" 
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+MahiFit::MahiFit() :
+  fullTSSize_(19), 
+  fullTSofInterest_(8)
+{}
+
+double MahiFit::getSiPMDarkCurrent(double darkCurrent, double fcByPE, double lambda) const {
+  double mu = darkCurrent * 25 / fcByPE;
+  return sqrt(mu/pow(1-lambda,3)) * fcByPE;
+}
+
+void MahiFit::setParameters(bool iDynamicPed, double iTS4Thresh, double chiSqSwitch, 
+			    bool iApplyTimeSlew, HcalTimeSlew::BiasSetting slewFlavor,
+			    double iMeanTime, double iTimeSigmaHPD, double iTimeSigmaSiPM, 
+			    const std::vector <int> &iActiveBXs, int iNMaxItersMin, int iNMaxItersNNLS,
+			    double iDeltaChiSqThresh, double iNnlsThresh) {
+
+  dynamicPed_    = iDynamicPed;
+
+  ts4Thresh_     = iTS4Thresh;
+  chiSqSwitch_   = chiSqSwitch;
+
+  applyTimeSlew_ = iApplyTimeSlew;
+  slewFlavor_    = slewFlavor;
+
+  meanTime_      = iMeanTime;
+  timeSigmaHPD_  = iTimeSigmaHPD;
+  timeSigmaSiPM_ = iTimeSigmaSiPM;
+
+  activeBXs_ = iActiveBXs;
+
+  nMaxItersMin_  = iNMaxItersMin;
+  nMaxItersNNLS_ = iNMaxItersNNLS;
+
+  deltaChiSqThresh_ = iDeltaChiSqThresh;
+  nnlsThresh_    = iNnlsThresh;
+
+  bxOffsetConf_ = -(*std::min_element(activeBXs_.begin(), activeBXs_.end()));
+  bxSizeConf_   = activeBXs_.size();
+
+}
+
+void MahiFit::phase1Apply(const HBHEChannelInfo& channelData,
+			  float& reconstructedEnergy,
+			  float& reconstructedTime,
+			  bool& useTriple, 
+			  float& chi2) const {
+
+  assert(channelData.nSamples()==8||channelData.nSamples()==10);
```

Comment:
why have this check given the recent changes towards a flexible # of samples?

Reply:
The two proposed number of samples are either 10 (as in Run 1 and 2015-2017 Run 2) and 8 (for 2018 Run 2 and foward). The flexible number of samples have all been in service of going from 10 to 8. (afaik)